CRAYSAT-1193:Improve create and delete fas snapshot

annapoorna-s-alt opened the pull request:
IM: CRAYSAT-1193
Reviewer: Ryan

Improved snapshot creation by removing expiration time 
Add delete option to prompt user to delete the snapshot if it is no longer needed.

## Summary and Scope

- Polling the snapshot in `sat firmware` resulted in HTTP errors in large clusters.
  Hence, adding the retry option when it consecutively fails for up to 5 times.
- Remove expiration time and add `--delete-snapshot` option to prompt the user to delete the snapshot
  if it is no longer needed. By default, log a message referring the user how to delete the snapshot.

## Issues and Related PRs

_Resolves [CRAYSAT-1193](https://jira-pro.it.hpe.com:8443/browse/CRAYSAT-1193)._

## Testing

_List the environments in which these changes were tested._

### Tested on:

  Baldar

### Test description:

_Test by creating a new snapshot using `sat firmware` command_
_Test by deleting the snapshot using `sat firmware --delete-snapshot <snapshote-name>`

## Risks and Mitigations

_Minimal_


## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

annapoorna-s-alt commented:
Please find the detailed output here: 
[CRAYSAT-1193.txt](https://github.com/Cray-HPE/sat/files/15108529/CRAYSAT-1193.txt)

haasken-hpe commented on sat/apiclient/fas.py:
```diff
@@ -339,13 +353,21 @@ def _create_temporary_snapshot(self, xnames=None):
-        while not ready:
-            time.sleep(1)
+        consecutive_failures = 0
+        while not ready and consecutive_failures < 5:
```
It looks like the check on `consecutive_failures` is unnecessary because it's handled any time it's incremented on line 365 below.

annapoorna-s-alt replied:
Got it. Removed on line 357.

haasken-hpe commented on sat/apiclient/fas.py:
```diff
@@ -354,6 +376,8 @@ def _create_temporary_snapshot(self, xnames=None):
             except KeyError:
                 raise APIError('Payload returned from GET to snapshots/name did not have "ready" field.')
 
+        LOGGER.info(f'Use `sat firmware --delete-snapshot` command to delete the snapshot if it is no longer needed')
```
I think you can drop "command" from this message.

```suggestion
        LOGGER.info(f'Use `sat firmware --delete-snapshot` to delete the snapshot if it is no longer needed')
```

haasken-hpe commented on sat/cli/firmware/main.py:
```diff
@@ -36,6 +36,24 @@
+        snapshot (str): The name of the snapshot to be deleted.
+
+    Raises:
+            APIError: If there is an error while attempting to delete the snapshot.
```
```suggestion
        SystemExit: If there is an error while attempting to delete the snapshot.
```

haasken-hpe commented on sat/cli/firmware/parser.py:
```diff
@@ -53,3 +53,8 @@ def add_firmware_subparser(subparsers):
         help='Describe firmware snapshots. Provide no arguments to this '
              'option to list available snapshots.')
+
+    firmware_parser.add_argument(
```
You might want to look into a way to make this mutually exclusive with the `--snapshots` option.

I don't really like how this command was structured to begin with, namely, I don't like that `sat firmware --snapshots` just lists existing snapshots and `sat firmware --snapshots SNAPSHOT_1 [SNAPSHOT_2 ...]` will show the firmware versions in those snapshots. I think it will be hard to change that behavior in a backwards compatible way.

It would just be nice to avoid any confusion in case somebody thinks that `sat firmware --delete-snapshot` means that it will create a snapshot, show the versions and then delete the snapshot. It might also be confusing that `sat firmware --delete-snapshot SNAPSHOT_1 --snapshots SNAPSHOT_1` will just delete SNAPSHOT_1 and will not show its contents. Perhaps you can think of a better way to structure the command line arguments.

annapoorna-s-alt replied:
I made `--snapshots` and `--delete-snapshot` as mutually exclusive. So user might not get confused here as they can use only one at a time. So  `sat firmware --delete-snapshot SNAPSHOT_1 --snapshots SNAPSHOT_1` would be invalid.

To be more specific `sat firmware --delete-snapshot SNAPSHOT_1` deletes the snapshot SNAPSHOT_1 if it is valid snapshot-name else throws an error.

haasken-hpe commented on sat/apiclient/fas.py:
```diff
@@ -324,8 +339,7 @@ def _create_temporary_snapshot(self, xnames=None):
 
-        # create a system snapshot set to expire in 10 minutes
-        payload = {'name': name, 'expirationTime': later.isoformat() + 'Z'}
+        payload = {'name': name}
```
the `later` variable is no longer used, which might be a little confusing.

One idea might be to rename this method  to `_create_snapshot` and have it take a boolean argument `expires` or `temporary` that controls whether the snapshot expires. Another idea could be to just rename it to `_create_snapshot` and get rid of the expiration logic and the `later` variable entirely.

annapoorna-s-alt replied:
I prefer to remove the expiration logic completely as it is nowhere used and renaming the function to `create_snapshot`

haasken-hpe reviewed (approved):
This looks good, nice job. I have some comments about usability for you to consider and if you want to implement any of those, I will re-review.

An additional usability question I have is whether it's desirable to be able to run `sat firmware --delete-snapshot` and have it just automatically delete the snapshot for you when it's done. Just an idea to consider.

haasken-hpe commented:
One thing you could try to do to test the resiliency behavior you added would be to start a `sat firmware` command on a decently large system and then between when `sat firmware` issues the POST request to create the snapshot and when it starts polling, perhaps you could force FAS into a bad state, possibly with a rollout restart or something in Kubernetes. The HMS folks could probably help you find a way to test that behavior.

annapoorna-s-alt commented:
### Tested on Odin.
Please find the attached output after addressing the comments.
[CRAYSAT-1193-final.txt](https://github.com/Cray-HPE/sat/files/15148783/CRAYSAT-1193-final.txt)

haasken-hpe commented on sat/apiclient/fas.py:
```diff
@@ -319,13 +333,12 @@ def _create_temporary_snapshot(self, xnames=None):
                     - The FAS did not indicate if the snapshot was ready.
         """
-        now, later = _now_and_later(10)
+        now = create_snapshot()
```
just remove `create_snapshot` and put the code that gets the date here

```suggestion
        now = datetime.utcnow().replace(microsecond=0)
```

haasken-hpe commented on sat/apiclient/fas.py:
```diff
@@ -339,13 +352,21 @@ def _create_temporary_snapshot(self, xnames=None):
+                    raise APIError('Error when polling the snapshot {}: Exceeded maximum consecutive failures',
+                                   format(err))
+                else:
+                    continue
```
I would suggest logging a debug-level message here.

annapoorna-s-alt replied:
Done

haasken-hpe commented on sat/apiclient/fas.py:
```diff
@@ -354,6 +375,8 @@ def _create_temporary_snapshot(self, xnames=None):
             except KeyError:
                 raise APIError('Payload returned from GET to snapshots/name did not have "ready" field.')
 
+        LOGGER.info(f'Use `sat firmware --delete-snapshot` to delete the snapshot if it is no longer needed')
```
Suggest moving this log message to occur after the table is output so the admin will see it.

annapoorna-s-alt replied:
Addressed

annapoorna-s-alt commented:
### Tested on Odin.

Please find the attached output after adding the additional feature to delete more than one snapshot
[CRAYSAT-1193-firmware.txt](https://github.com/Cray-HPE/sat/files/15187241/CRAYSAT-1193-firmware.txt)

haasken-hpe reviewed (approved):
Looks good!
